session_test.go

```diff
@@ -0,0 +1,313 @@
+
+	if err := session.Close(); err != nil {
+		p.Fail("Failed to close session properly; %s", err)
+	}
+
+	// Ensure processing goroutine wont block forever.
+	select {
+	case <-done:
+		// Pass.
+	case <-time.After(deadline):
+		p.Fail("Failed to stop event processing in 10s")
+	}
+}
+
+func (p *testProvider) TestUpdating() {
+	ctx, cancel := context.WithCancel(context.Background())
+	go p.generateEvents(ctx, []etw.Level{etw.LevelInfo, etw.LevelCritical})
+	defer func() {
+		cancel()
+	}()
+
+	session, err := NewSession(p.guid, WithLevel(TRACE_LEVEL_CRITICAL))
+	if err != nil {
+		p.Require().NoError(err, "Failed to create session")
+	}
+
+	gotCriticalEvent := make(chan struct{}, 1)
+	gotInformationEvent := make(chan struct{}, 1)
+	cb := func(e *Event) {
+		switch TraceLevel(e.Header.Level) {
+		case TRACE_LEVEL_INFORMATION:
+			select {
+			case gotInformationEvent <- struct{}{}:
+			default:
+			}
+		case TRACE_LEVEL_CRITICAL:
+			select {
+			case gotCriticalEvent <- struct{}{}:
+			default:
+			}
+		default:
+		}
+	}
+
+	done := make(chan struct{})
+	go func() {
+		err := session.Process(cb)
+		p.Require().NoError(err, "Failed to start processing events")
+		close(done)
+	}()
+
+	deadline := 10 * time.Second
+
+	// Ensure that we are consuming events with the needed level.
+	go func() {
+		select {
+		case <-gotCriticalEvent:
+			// pass
+		case <-time.After(deadline):
+			p.Fail("Failed to get event with critical level")
+		}
+	}()
+
+	// Ensure that we aren't consuming events with unnecessary levels.
+	select {
+	case <-gotInformationEvent:
+		p.Fail("Received event with unexpected level")
+	case <-time.After(deadline):
+		// pass
+	}
+
+	if err := session.UpdateOptions(WithLevel(TRACE_LEVEL_INFORMATION)); err != nil {
+		p.Fail("Failed to update session options; %s", err)
+	}
+
+	// Ensure option updates are applied.
+	select {
+	case <-gotInformationEvent:
+		// pass
+	case <-time.After(deadline):
+		p.Fail("Failed to receive event with INFO level after updating session options")
+	}
+
+	// Stop the session properly.
+	if err := session.Close(); err != nil {
+		p.Fail("Failed to close session properly; %s", err)
+	}
+	// Ensure processing goroutine wont block forever.
+	select {
+	case <-done:
+		// Pass.
+		close(gotCriticalEvent)
+		close(gotInformationEvent)
+	case <-time.After(deadline):
+		p.Fail("Failed to stop event processing in 10s")
+	}
+}
+
+func (p *testProvider) TestParsing() {
+	ctx, cancel := context.WithCancel(context.Background())
+	go p.generateEvents(
+		ctx,
+		[]etw.Level{etw.LevelInfo},
+		etw.StringField("string", "string value"),
+		etw.StringArray("stringArray", []string{"string1", "string2", "string3"}),
+		etw.Float64Field("float64", 45.7),
+		etw.Struct("struct",
+			etw.StringField("string", "string value"),
+			etw.StringArray("stringArray", []string{"string1", "string2", "string3"}),
+			etw.Float64Field("float64", 46.7),
+			etw.Struct("subStructure",
+				etw.StringField("string", "string value"))),
+	)
```

Comment:
How about array of structs? :)

Reply:
Microsoft etw doesn't have function `StructArray`  :(